Upgrade apache-libcloud package dependency for 2.0

tonybaloney opened the pull request:
### What does this PR do?

This package updates the apache-libcloud dependency for salt-cloud to the 2.0 version, from 0.14. 

0.14 is a _very_ old version of libcloud and there have been performance improvements, stability changes and security fixes in the past 2 years that salt-cloud could take advantage of. Most of the salt-cloud modules that use libcloud e.g. GCP and Dimension Data require a newer version anyway, so the user just has to upgrade. 

The 2.0 release also no longer uses httplib, now uses the `requests` package, which is already a dependency of salt and doesn't require SSL cert configuration for the CA store. The implications are 
http://libcloud.readthedocs.io/en/latest/other/changes_in_2_0.html
2.0 supports Python 3.6, drops 2.5 and 3.2. Python 2.6 support is still included.

This is an open PR for discussion, look at the downstream dependencies and their impact. Also, part of a conversation with @techhat 

Please review [Salt's Contributing Guide](https://docs.saltstack.com/en/latest/topics/development/contributing.html) for best practices.

I've left a comment over the 2.0 build as only rc2 is on PyPi at the moment but the full release comes out this week.

tonybaloney committed:
upgrade apache-libcloud packages for 2.0

salt-jenkins commented:
@tonybaloney, thanks for your PR! By analyzing the history of the files in this pull request, we identified @vutny, @twangboy and @s0undt3ch to be potential reviewers.

tonybaloney commented:
i.e. don't merge yet, please :-)

tonybaloney committed:
remove workarounds for #32744 and also references to the certifi certificate pack

tonybaloney commented:
This commit is because libcloud doesn't require a CA cert bundle anymore. https://github.com/saltstack/salt/pull/40837/commits/399f2a783d7f422279fb663b8a87436ebbf389ef The referenced issue is a workaround to load the CA bundle for SuSE machines to the YaST PEM path. 
This is no longer needed because `requests` includes a CA bundle

cachedout commented:
Hi @tonybaloney !

A test was unhappy here as well: https://jenkins.saltstack.com/job/PR/job/salt-pr-rs-cent7-n/10755/testReport/junit/integration.modules.test_libcloud_dns/LibcloudDNSTest/test_list_record_types/

tonybaloney committed:
add some assertion messages to help debug the issue

tonybaloney committed:
remove trailing whitespace

tonybaloney commented:
this PR can go on hold until Friday 

```
Traceback (most recent call last):
  File "/testing/tests/integration/modules/test_libcloud_dns.py", line 19, in test_list_record_types
    self.assertTrue('SPF' in supported_records, msg="Failed to find record type 'SPF' in {0}".format(supported_records))
AssertionError: Failed to find record type 'SPF' in 'libcloud_dns' __virtual__ returned False: A apache-libcloud library with version at least 2.0.0 was not found
```
It's because the workaround I put in before the package goes to PyPi isn't working properly, https://github.com/saltstack/salt/pull/40837/files#diff-8e6587eeb96caa87567943f6699715b4R5

tonybaloney committed:
Merge branch 'develop' into libcloud_2

tonybaloney committed:
move to the pypi release of 2.0

tonybaloney committed:
remove the integration test for dns

tonybaloney commented:
the pypi release is out. The integration test still fails. I think whatever script is used to deploy the VMs isn't installing the right version. So I've removed the integration test anyway (and added far more coverage in other unit tests).

tonybaloney commented:
so https://github.com/saltstack/salt/pull/40837/commits/1ff7ee302aee3a8e6d074a9e9e9553a3a9a47e20 should go green this this can be merged.

vutny commented on salt/cloud/deploy/bootstrap-salt.sh:
```diff
@@ -227,7 +227,7 @@ _SALT_MASTER_ADDRESS=${BS_SALT_MASTER_ADDRESS:-null}
 # _SIMPLIFY_VERSION is mostly used in Solaris based distributions
 _SIMPLIFY_VERSION=$BS_TRUE
-_LIBCLOUD_MIN_VERSION="0.14.0"
+_LIBCLOUD_MIN_VERSION="2.0.0"
```
I wonder if this change should go to https://github.com/saltstack/salt-bootstrap as well.
However, it will affect only `pip`-enabled installations...

tonybaloney replied:
I honestly don't know who updates the the repos. But I could certainly help. I'll look at salt-bootstrap. Something is causing the version errors

vutny replied:
I think that's the question for @dmurphy18.

That line in `bootstrap-salt.sh` actually only makes sense when you install `salt-cloud` itself with `pip` using the script, not `salt-minion`. So, that's not a big deal though.
I have brought attention to it, because I know this particular script is just copied from [salt-bootstrap](https://github.com/saltstack/salt-bootstrap) repo during stable releases.
There is a high chance that your change will be just overwritten :smile:

cachedout replied:
If it's bootstrap, we want to talk to @rallytime here. :]

rallytime replied:
@vutny Is right. If changes are being made to the bootstrap script in salt, they will just be overwritten when we release a new bootstrap script. And changes to `bootstrap-salt.sh` need to be made in the [salt-bootstrap](https://github.com/saltstack/salt-bootstrap) repo if they are to persist after the next bootstrap version is released.

vutny reviewed (commented):
Are there any plans to package new version of `python-libcloud` and provide it from SaltStack's corp repo?

I assume that depending on `2.0.0` is not strict, so most `salt-cloud` functionality should still work with previous versions, right?

tonybaloney commented:
Looks like someone at redhat updates the packages. http://pkgs.fedoraproject.org/cgit/rpms/python-libcloud.git

cachedout commented:
I just want @dmurphy18 to sign off on this before we merge it. LGTM though.

tonybaloney commented:
Fwiw I found the package source for the fedora packages and what I assume is being used in yum. http://pkgs.fedoraproject.org/cgit/rpms/python-libcloud.git

dmurphy18 commented:
@cachedout @techhat  Reached out to @techhat yesterday waiting for reply.
is there a new minimum version of libcloud that I should be packaging as a dependency, typically I package an old minion (varies by platform) and let the platforms version override if newer, for example: only provide a basic version for those insane people who run without EPEL

I don't believe the minimum version we need has changed, but awaiting @techhat 's response

techhat commented:
I would prefer to update this dependency, as libcloud 2.0 is much more stable, and upgrading it now will make work with certain drivers going forward easier. But I don't know of a absolute, specific requirement for the upgrade now. But if we skip it now, you can expect it will be a hard requirement by the time the next major version comes out.

dmurphy18 reviewed (commented):
@cachedout @techhat @vutny @tonybaloney  I vote this down since this minimum requirement of v2.0.0 will break platforms which require packaged versions.  For example: current version of python-libcloud for Redhat 7 from EPEL is:
root@localhost:~# yum list python-libcloud
Loaded plugins: fastestmirror, langpacks
Loading mirror speeds from cached hostfile
 * base: repo1.sea.innoscale.net
 * epel: mirror.sfo12.us.leaseweb.net
 * extras: mirrors.xmission.com
 * updates: mirror.keystealth.org
Available Packages
python-libcloud.noarch                                                                               0.20.1-1.el7                                                                                epel
root@localhost:~# 

Similar version 0.20 for Ubuntu 16.04, and Debian 7 is even older.

The only version of 2.0.0 I could find for the Redhat family is Fedora 26 (Alpha) which is in testing.

Before implementing this change SaltStack would need to provide v2.0.0 packages for all platforms we support.

cachedout commented:
@tonybaloney Since you're in tight with the libcloud folks..any insight on the state of packaging for those distros? If we have to do it ourselves, it might take a while. :]

toanju commented:
@dmurphy18 Fedora has updated libcloud on F24 and F25: https://apps.fedoraproject.org/packages/python-libcloud 

They (unfortuneately) arrived yesterday:
https://bodhi.fedoraproject.org/updates/FEDORA-2017-aeda7ad8c8 (direct link to F24)

dmurphy18 commented:
@toanju I think it would be better to wait till they are no longer Release Candidate 2. Once they are no longer a Release Candidate and actually a released version, I shall review building that version.

toanju commented:
@dmurphy18 in general I would agree, but a released package in this case is imho a release regardless of the name of the package. The package is considered stable in terms of Fedora and for everyone salt-cloud is currently broken.

cachedout commented:
@toanju Could you please provide some specifics on how it is broken?

toanju commented:
@cachedout besides the patch I made in #41066, using salt-cloud 2016.11.4 (Carbon) with libcloud 2.0.0rc2 I am running into 

```
[root@salt ~]# salt-cloud -Q
[ERROR   ] Failed to import clouds cloudstack, this is due most likely to a syntax error:
Traceback (most recent call last):
  File "/usr/lib/python2.7/site-packages/salt/loader.py", line 1333, in _load_module
    mod = imp.load_module(mod_namespace, fn_, fpath, desc)
  File "/usr/lib/python2.7/site-packages/salt/cloud/clouds/cloudstack.py", line 43, in <module>
    libcloud.security.CA_CERTS_PATH.append('/etc/ssl/certs/YaST-CA.pem')
AttributeError: 'NoneType' object has no attribute 'append'
[ERROR   ] Failed to import clouds rackspace, this is due most likely to a syntax error:
Traceback (most recent call last):
  File "/usr/lib/python2.7/site-packages/salt/loader.py", line 1333, in _load_module
    mod = imp.load_module(mod_namespace, fn_, fpath, desc)
  File "/usr/lib/python2.7/site-packages/salt/cloud/clouds/rackspace.py", line 60, in <module>
    libcloud.security.CA_CERTS_PATH.append('/etc/ssl/certs/YaST-CA.pem')
AttributeError: 'NoneType' object has no attribute 'append'
[WARNING ] /usr/lib/python2.7/site-packages/salt/cloud/clouds/openstack.py:229: DeprecationWarning: This driver has been deprecated and will be removed in the Oxygen (Unreleased) release of Salt. Please use the nova driver instead.

[ERROR   ] Failed to import clouds dimensiondata, this is due most likely to a syntax error:
Traceback (most recent call last):
  File "/usr/lib/python2.7/site-packages/salt/loader.py", line 1333, in _load_module
    mod = imp.load_module(mod_namespace, fn_, fpath, desc)
  File "/usr/lib/python2.7/site-packages/salt/cloud/clouds/dimensiondata.py", line 40, in <module>
    libcloud.security.CA_CERTS_PATH.append('/etc/ssl/certs/YaST-CA.pem')
AttributeError: 'NoneType' object has no attribute 'append'
[ERROR   ] Failed to import clouds gce, this is due most likely to a syntax error:
Traceback (most recent call last):
  File "/usr/lib/python2.7/site-packages/salt/loader.py", line 1333, in _load_module
    mod = imp.load_module(mod_namespace, fn_, fpath, desc)
  File "/usr/lib/python2.7/site-packages/salt/cloud/clouds/gce.py", line 70, in <module>
    libcloud.security.CA_CERTS_PATH.append('/etc/ssl/certs/YaST-CA.pem')
AttributeError: 'NoneType' object has no attribute 'append'
[ERROR   ] Failed to import clouds cloudstack, this is due most likely to a syntax error:
Traceback (most recent call last):
  File "/usr/lib/python2.7/site-packages/salt/loader.py", line 1333, in _load_module
    mod = imp.load_module(mod_namespace, fn_, fpath, desc)
  File "/usr/lib/python2.7/site-packages/salt/cloud/clouds/cloudstack.py", line 43, in <module>
    libcloud.security.CA_CERTS_PATH.append('/etc/ssl/certs/YaST-CA.pem')
AttributeError: 'NoneType' object has no attribute 'append'
[ERROR   ] Failed to import clouds rackspace, this is due most likely to a syntax error:
Traceback (most recent call last):
  File "/usr/lib/python2.7/site-packages/salt/loader.py", line 1333, in _load_module
    mod = imp.load_module(mod_namespace, fn_, fpath, desc)
  File "/usr/lib/python2.7/site-packages/salt/cloud/clouds/rackspace.py", line 60, in <module>
    libcloud.security.CA_CERTS_PATH.append('/etc/ssl/certs/YaST-CA.pem')
AttributeError: 'NoneType' object has no attribute 'append'
[ERROR   ] Failed to import clouds cloudstack, this is due most likely to a syntax error:
Traceback (most recent call last):
  File "/usr/lib/python2.7/site-packages/salt/loader.py", line 1333, in _load_module
    mod = imp.load_module(mod_namespace, fn_, fpath, desc)
  File "/usr/lib/python2.7/site-packages/salt/cloud/clouds/cloudstack.py", line 43, in <module>
    libcloud.security.CA_CERTS_PATH.append('/etc/ssl/certs/YaST-CA.pem')
AttributeError: 'NoneType' object has no attribute 'append'
[ERROR   ] Failed to import clouds dimensiondata, this is due most likely to a syntax error:
Traceback (most recent call last):
  File "/usr/lib/python2.7/site-packages/salt/loader.py", line 1333, in _load_module
    mod = imp.load_module(mod_namespace, fn_, fpath, desc)
  File "/usr/lib/python2.7/site-packages/salt/cloud/clouds/dimensiondata.py", line 40, in <module>
    libcloud.security.CA_CERTS_PATH.append('/etc/ssl/certs/YaST-CA.pem')
AttributeError: 'NoneType' object has no attribute 'append'
[ERROR   ] Failed to import clouds rackspace, this is due most likely to a syntax error:
Traceback (most recent call last):
  File "/usr/lib/python2.7/site-packages/salt/loader.py", line 1333, in _load_module
    mod = imp.load_module(mod_namespace, fn_, fpath, desc)
  File "/usr/lib/python2.7/site-packages/salt/cloud/clouds/rackspace.py", line 60, in <module>
    libcloud.security.CA_CERTS_PATH.append('/etc/ssl/certs/YaST-CA.pem')
AttributeError: 'NoneType' object has no attribute 'append'
[ERROR   ] Failed to import clouds gce, this is due most likely to a syntax error:
Traceback (most recent call last):
  File "/usr/lib/python2.7/site-packages/salt/loader.py", line 1333, in _load_module
    mod = imp.load_module(mod_namespace, fn_, fpath, desc)
  File "/usr/lib/python2.7/site-packages/salt/cloud/clouds/gce.py", line 70, in <module>
    libcloud.security.CA_CERTS_PATH.append('/etc/ssl/certs/YaST-CA.pem')
AttributeError: 'NoneType' object has no attribute 'append'
[ERROR   ] Failed to import clouds cloudstack, this is due most likely to a syntax error:
Traceback (most recent call last):
  File "/usr/lib/python2.7/site-packages/salt/loader.py", line 1333, in _load_module
    mod = imp.load_module(mod_namespace, fn_, fpath, desc)
  File "/usr/lib/python2.7/site-packages/salt/cloud/clouds/cloudstack.py", line 43, in <module>
    libcloud.security.CA_CERTS_PATH.append('/etc/ssl/certs/YaST-CA.pem')
AttributeError: 'NoneType' object has no attribute 'append'
[ERROR   ] Failed to import clouds rackspace, this is due most likely to a syntax error:
Traceback (most recent call last):
  File "/usr/lib/python2.7/site-packages/salt/loader.py", line 1333, in _load_module
    mod = imp.load_module(mod_namespace, fn_, fpath, desc)
  File "/usr/lib/python2.7/site-packages/salt/cloud/clouds/rackspace.py", line 60, in <module>
    libcloud.security.CA_CERTS_PATH.append('/etc/ssl/certs/YaST-CA.pem')
AttributeError: 'NoneType' object has no attribute 'append'
[ERROR   ] Failed to import clouds dimensiondata, this is due most likely to a syntax error:
Traceback (most recent call last):
  File "/usr/lib/python2.7/site-packages/salt/loader.py", line 1333, in _load_module
    mod = imp.load_module(mod_namespace, fn_, fpath, desc)
  File "/usr/lib/python2.7/site-packages/salt/cloud/clouds/dimensiondata.py", line 40, in <module>
    libcloud.security.CA_CERTS_PATH.append('/etc/ssl/certs/YaST-CA.pem')
AttributeError: 'NoneType' object has no attribute 'append'
[ERROR   ] Failed to import clouds dimensiondata, this is due most likely to a syntax error:
Traceback (most recent call last):
  File "/usr/lib/python2.7/site-packages/salt/loader.py", line 1333, in _load_module
    mod = imp.load_module(mod_namespace, fn_, fpath, desc)
  File "/usr/lib/python2.7/site-packages/salt/cloud/clouds/dimensiondata.py", line 40, in <module>
    libcloud.security.CA_CERTS_PATH.append('/etc/ssl/certs/YaST-CA.pem')
AttributeError: 'NoneType' object has no attribute 'append'
[ERROR   ] Failed to import clouds gce, this is due most likely to a syntax error:
Traceback (most recent call last):
  File "/usr/lib/python2.7/site-packages/salt/loader.py", line 1333, in _load_module
    mod = imp.load_module(mod_namespace, fn_, fpath, desc)
  File "/usr/lib/python2.7/site-packages/salt/cloud/clouds/gce.py", line 70, in <module>
    libcloud.security.CA_CERTS_PATH.append('/etc/ssl/certs/YaST-CA.pem')
AttributeError: 'NoneType' object has no attribute 'append'
[ERROR   ] Failed to import clouds gce, this is due most likely to a syntax error:
Traceback (most recent call last):
  File "/usr/lib/python2.7/site-packages/salt/loader.py", line 1333, in _load_module
    mod = imp.load_module(mod_namespace, fn_, fpath, desc)
  File "/usr/lib/python2.7/site-packages/salt/cloud/clouds/gce.py", line 70, in <module>
    libcloud.security.CA_CERTS_PATH.append('/etc/ssl/certs/YaST-CA.pem')
AttributeError: 'NoneType' object has no attribute 'append'
```
locally I have worked around this by simply removing the calls to `libcloud.security.CA_CERTS_PATH.append`. The nicer solution might check the version though.

cachedout commented:
@toanju Would you mind filing that stacktrace as another issue? My feeling here is that even if we do upgrade libcloud, we need to adjust those calls that are producing stacktraces to be more defensive anyhow.

toanju commented:
@cachedout I spent some more time on this, it turns out that I had a configuration error and I got confused by the errors shown above. Hence sorry for the noise.

tonybaloney commented:
Those lines are totally removed in this PR, it's a workaround to make sure that even for older version of liblcoud (<0.20.0) it works smoothly in SuSE. 
2.0 doesn't need CA paths or bundles. 

I think I'll remove the changes to the bootstrap script and this PR will only serve PIP-managed installations. In parallel, I then need to find out how to update those packages!

http://libcloud.readthedocs.io/en/latest/changelog.html#changes-in-apache-libcloud-2-0-0

cachedout commented:
@tonybaloney Can you comment on the concern from @dmurphy18 above on going with an RC instead of waiting for a full release? (We also need a rebase here I'm afraid since this now has merge conflicts.)

tonybaloney committed:
Merge branch 'develop' into libcloud_2

tonybaloney committed:
revert bootstrap change

tonybaloney commented:
Reverted the bootstrap change so this is just for pip-managed installations.

I've also spoken to the Fedora package maintainers, fc 24,25 and 26 are getting 2.0.0 (release, not rc2) packages now e.g. https://koji.fedoraproject.org/koji/buildinfo?buildID=887856

I have a broader question here as to why we should install libcloud from deb/rpm? @dmurphy18 

We need to get to the 2.0 series eventually in the various package repos

dmurphy18 commented:
@tonybaloney I am a little confused by your question: 

I have a broader question here as to why we should install libcloud from deb/rpm? @dmurphy18

SaltStack provides dependencies packaged at a minimum level for Salt to function, with the aim that a  user could override the version provided by SaltStack when a newer becomes available, for example by EPEL or later version of Debian 8 - some of the dependencies did not exist in 8.0, hence their inclusion.

Providing a later version of a package which could override the commonly used version for the platform may not be in the best interests of the user: for example: libcloud versions for Debian 8.5 is 0.15.1-1 and Debian 7.11 is 0.5.0-1, hence SaltStack provides a minimum version of 0.20.0-1. On Ubuntu 14.04 SaltStack provides libcloud 0.15.1-1 which is now available from trusty-backports but SaltStack still provides that version in case a user does not have backports enabled.

It might not be the best practice to update the packages and override the versions provided by the platform, unless there was a compelling need.  However pip installing an updated libcloud would be a valid practice since that is at the user's discretion.

Please let me know if I have answered your question satisfactorily.

tonybaloney commented:
@dmurphy18 libcloud by itself doesn't do anything unless you instantiate the drivers, of which are continually changing. 
They have to change because the cloud APIs they called are being updated monthly. GCE as an example looks very different in 2017 to 2014 (when 0.14 was released). 

I'm more concerned that the user will have libcloud installed, but the code that depends on it won't work. Also, all of the cloud drivers depend on newer versions, GCE needs minimum 0.21 and I would strongly recommend using 1.0 as a minimum. 

"It might not be the best practice to update the packages and override the versions provided by the platform, unless there was a compelling need. However pip installing an updated libcloud would be a valid practice since that is at the user's discretion."

Wondering if we can just install via Pip anyway, so it's always the latest stable or just pin a particular release. I'd also never install requests via yum as another example, because the CA root store is bundled into the package and you want the latest for security purposes.

cachedout commented:
@dmurphy18 Could you please come back to this one? Thanks. :]

dmurphy18 commented:
@techhat Is there a new minimum version of libcloud that you recommend that we package, currently using 0.20.0 ?

techhat commented:
@erjohnso do you have any thoughts on this?

erjohnso commented:
Putting on my Google hat (it's never far away), I'm definitely in favor of upping to libcloud 2.0 as soon as possible. Anthony has done a **ton** of work on libcloud across many drivers, and distro updates, so if we want Salt to take advantage of that, we need to get this upgraded sooner than later.

Regarding pip vs distro packages, I think moving to a known pip version likely makes the most sense. I don't know that there's a perfect answer for avoiding versioning issues between upstream, pypi, distro packages - but I'm inclined to think that using pip allows you to stay more current and only risking breakage on very old distros.

techhat commented:
@erjohnso, we have a need for OS-level packaging to be in place. Do you have a suggestion on a minimum version that is already packaged for the major distros?

erjohnso commented:
Hmm... No, my distro-hopping days are ancient history. Here's the [changelog](https://libcloud.readthedocs.io/en/latest/changelog.html) to give a sense of how many releases have been published (and features added) since 0.20.0.

dmurphy18 commented:
I'll look into doing 2.0 versions of libcloud for Nitrogen packages then.

dmurphy18 commented:
Note given the age and soon to be released Debian 9, not planning on upgrading the current version supplied for Debian 7  (0.20.0), since it will get pushed from oldstable once Debian 9 is released.

cachedout commented:
So, in looking at the comments here: https://github.com/saltstack/salt-pack/issues/314

One question I have is whether we need to go all the way to 2.0 or whether 1.5.0 is sufficient.

cc: @dmurphy18

cachedout commented:
@tonybaloney Do you have any sway with the Debian/Ubuntu maintainers of the libcloud packages? Any way we can know when/if they're going to package 2.x?

cachedout commented:
Bump @tonybaloney. Sorry to bug you, but did you see the question above by chance?

tonybaloney commented:
@cachedout I made progress with centos/rhel but not with debian. I have the ML details now.

cachedout commented:
@tonybaloney Cool. Keep us posted. :]

dmurphy18 commented:
@tonybaloney FYI - Redhat family (7, 6 Amazon) for Nitrogen shall have python-libcloud v2.0.0.  Debian 9 & 8, Ubuntu 16.04 & 14.04 shallhave python-libcloud v1.5.0.  The Debian / Ubuntu families are using v1.5.0 since that is the latest released version of python-libcloud available for those OS platforms.

In the upcoming point release Debian 8 and Raspbian (Debian 8 armhf) python-libcloud shall be v1.5.0: this is due to build issue with the older v0.20.0 and it was easier to just use the version 1.5.0 that built cleanly which was developed for Nitrogen.

rallytime commented:
Hey @tonybaloney I was just swinging back around here and it looks like this has merge conflicts now. When you get a second, can you rebase this and resolve them? Thanks again for your work here.

rallytime commented:
bump @tonybaloney  :)

rallytime commented:
Since we haven't heard back here for a while, I am going to close this. We can of course open it again when you're ready to come back to it. There's several merge conflicts here though, so it will need to be rebased against current develop and the merge conflicts resolved. :) Thank you!
